File: examples/governance/compliance_domain/lib/compliance_domain/compliance_review/commands/open_review.rb
module ComplianceDomain
  class ComplianceReview
    module Commands
      class OpenReview
        include Hecks::Command
        emits "OpenedReview"

        attr_reader :model_id
        attr_reader :policy_id
        attr_reader :reviewer_id

        def initialize(
          model_id: nil,
          policy_id: nil,
          reviewer_id: nil
        )
          @model_id = model_id
          @policy_id = policy_id
          @reviewer_id = reviewer_id
        end

        def call
          ComplianceReview.new(
            model_id: model_id,
            policy_id: policy_id,
            reviewer_id: reviewer_id,
            status: "open"
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/open_review.go
package domain

import (
	"time"
)

type OpenReview struct {
	ModelId string `json:"model_id"`
	PolicyId string `json:"policy_id"`
	ReviewerId string `json:"reviewer_id"`
}

func (c OpenReview) CommandName() string { return "OpenReview" }

func (c OpenReview) Execute(repo ComplianceReviewRepository) (*ComplianceReview, *OpenedReview, error) {
	agg := NewComplianceReview(c.ModelId, c.PolicyId, c.ReviewerId, "", "", time.Time{}, nil, "")
	agg.Status = "open"
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := OpenedReview{
		AggregateID: agg.ID,
		ModelId: c.ModelId,
		PolicyId: c.PolicyId,
		ReviewerId: c.ReviewerId,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
